### PR TITLE
[df] Use maps for internal type lookups

### PR DESCRIPTION
**Pull request by vepadulano**
Alternative to #16529 @pcanal

### CODE REVIEW

File: tree/dataframe/src/RDFUtils.cxx [hunk 1]
```diff
@@ -40,6 +40,96 @@ ROOT::Experimental::RLogChannel &ROOT::Detail::RDF::RDFLogChannel()
...
+using TypeInfoRef = std::reference_wrapper<const std::type_info>;
+struct TypeInfoRefHash {
+   std::size_t operator()(TypeInfoRef id) const { return id.get().hash_code(); }
+};
+
+struct TypeInfoRefEqualComp {
+   bool operator()(TypeInfoRef left, TypeInfoRef right) const { return left.get() == right.get(); }
+};
+
+const static std::unordered_map<std::string, TypeInfoRef> gTypeName2TypeIDMap{
```
Comment (pcanal) [inline tree/dataframe/src/RDFUtils.cxx:53]: Note that declaring the array at the global scope (as opposed to the function scope) means:
* Initialization is done at library loading time in an order determine by both the order of libraries and the order of object file within the library
* If the using functions (i.e. `TypeName2TypeID` etc.) are called through a routine executed during library loading time (i.e. some other global static initialization) there is no guarantee that the map would be have been initialized at that time
* on the flip side, if the map are declared as function static, each call incurs a thread-safe checks that they have already been initialized.

In conclusion, the current choice is the right choice if and only if we are 100% sure that the functions will never be called (directly or indirectly) from a static initialization.

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: Thanks for your very valid input. Let me include some comments relevant to understand the situation better:
* The functions are in `Internal` namespace and they are used only within `tree/dataframe` source code
* The functions are inherently runtime-bound since they deal with the logic of finding a type for a column or a name for a column type, which only makes sense the user starts building the RDataFrame computation graph or they trigger it (e.g. the first time a user calls `Define`, the first time they trigger an action etc.).
* The following program which loads the library is not calling any of the functions touched in this PR
```cpp
#include <dlfcn.h>

int main()
{
    dlopen("/Users/vpadulan/Programs/rootproject/rootbuild/rdf-typeid-lookup-maps-distrdf-debug/lib/libROOTDataFrame.so", RTLD_NOW);
}
```
```
lldb ./rdf_dlopen.out                                                                                          130 ↵
(lldb) target create "./rdf_dlopen.out"
Current executable set to '/tmp/rdf_dlopen.out' (arm64).
(lldb) b ROOT::Internal::RDF::TypeName2ROOTTypeName
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) b ROOT::Internal::RDF::TypeName2TypeID
Breakpoint 2: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) b ROOT::Internal::RDF::TypeID2TypeName
Breakpoint 3: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
Process 11170 launched: '/tmp/rdf_dlopen.out' (arm64)
1 location added to breakpoint 1
1 location added to breakpoint 2
1 location added to breakpoint 3
Process 11170 exited with status = 0 (0x00000000)
```

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: That is true, what I meant in my comment is that they are not used anywhere else within ROOT's code. It is in principle conceivable to see a downstream project trying to force its way and use `ROOT::Internal` functions, which of course if we knew about we would highly discourage. But yes it's still possible.

My question is, considering the discussion we had at https://github.com/root-project/root/pull/16529, would you rather introduce the changes in that PR, or do you prefer the changes in this PR amended by putting the static maps at function scope?

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: > The following program which loads the library is not calling any of the functions touched in this PR

I would not expect it.

The problem would only arise in the rare (bordering shooting oneself on the foot).  where somebody write directly or indirectly:
```
static int amItoocleverformyowngood = []() {
    TFile file(myfilename);
    RDataFrame df;
    df .....
    return df.SomeCount();
}
```
I am not advocating to support this.  I am asking the 2 general questions:  How likely is it to happen (and get in the weird setup case)? (answer: unlikely).  How can we 'cheaply' make it fail elegantly?

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: > do you prefer the changes in this PR amended

I prefer the change in this PR.

> amended by putting the static maps at function scope?

I am not sure that the cost ( thread-safe check at each invocation ) is worth the benefit (working when called (indirectly via the official RDataFrame interface) from the static initialization)

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: On the other hand if those calls are only made during the construction of the `RDataFrame` tree of nodes then the cost of the thread-safe check might be minimal.

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: No, the calls are not made only during the building of the computation graph. These calls can and will also happen in-between multithreaded tasks. This is why https://github.com/root-project/root/pull/16528 was needed in the first place

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: Then I think we should keep the code in this PR essential as-is (using static globals).  We should also document (for good measure) that `RDataFrame` is not supported during static initialization and then add a (potential one time) check in the `RDataFrame` constructor that the static initialization that we need have been done. i.e.
```
if (R__unlikely(gIntWithSpecialValue != kIntSpecialValue))
   throw("Using RDataFrame during static initialization is not supported).
```

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: I don't understand your last point, what are `gIntWithSpecialValue, kIntSpecialValue` and where should they be defined?

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: Also, slightly unrelated, is `TFile` as well as any other core class completely safe to use in static initialization contexts?

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: To a large extent.  For example `gROOT` is actual a macro than turn into a (overly complex) set of calls to ends up in a function static initialization.  Again I am not advocating support for running code during initialization just adding a way to clearly fail.  The proposal boils down to having:
```
static /* must not be const */ size_t kMagicInitNumber = 0xdeadbeef; // really any number that is unlikely to be there in memory before initilization.
....
const static std::unordered_map<std::string, TypeInfoRef> gTypeName2TypeIDMap {
...
};
....
const static std::unordered_map<std::string, char> gTypeName2ROOTTypeNameMap {
...
};
// Very last initialization of the file (but probably does not make a real difference where in the file its since
// all the initialization will be in a row.
/* extern or somehow reacheable from the `RDataFrame` constructor
size_t gRDFUtilsHasBeenInitialized = kMagicInitNumber;
```

See the running example of such setup in 
[init.example.tar.gz](https://github.com/user-attachments/files/17211514/init.example.tar.gz)
and use with:
```
tar xfz init.example.tar.gz
cd init
./go
```

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: And with the above names, the `RDataFrame` constructors gets:
```
if (R__unlikely(gRDFUtilsHasBeenInitialized != kMagicInitNumber))
   throw("Using RDataFrame during static initialization is not supported).
```

Comment (vepadulano) [reply tree/dataframe/src/RDFUtils.cxx:53]: I can confirm that moving the definition of the maps to function-local `const static` does not change the performance of the AGC on a 128 core machine with IMT.

Comment (pcanal) [reply tree/dataframe/src/RDFUtils.cxx:53]: Ok.  That's the way to go then.

**Review by pcanal — APPROVED** [review]
Thanks.
